File: leetcode_learning/code_cpp/24.cpp
//Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 

//递归
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head==NULL||head->next==NULL)
            return head;
        ListNode* newhead = head->next;
        head->next = swapPairs(newhead->next);
        newhead->next = head;
        return newhead;
    }
};


//迭代
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* fake_head = new ListNode(0);
        fake_head->next = head;
        ListNode* temp = fake_head;
        while(temp->next!=NULL&&temp->next->next!=NULL)
        {
            ListNode* node1 = temp->next;
            ListNode* node2 = temp->next->next;
            temp->next = node2;
            node1->next = node2->next;
            node2->next = node1;
            temp = node1;
        }
        return fake_head->next;
    }
};

File: leetcode_learning/code_py/24.py

class ListNode():
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    # 递归方法
    # def swapPairs(self, head):
    #     """
    #     :type head: Optional[ListNode]
    #     :rtype: Optional[ListNode]
    #     """
    #     if (head is None) or (head.next is None):
    #         return head
    #     new_head = head.next
    #     head.next = self.swapPairs(new_head.next)
    #     new_head.next = head
    #     return new_head

    # 遍历
    def swapPairs(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if (head is None) or (head.next is None):
            return head
        fake_head = ListNode(0, head)
        cur = fake_head
        while(cur.next and cur.next.next):
            node1 = cur.next
            node2 = cur.next.next
            cur.next = node2
            node1.next = node2.next
            node2.next = node1
            cur = node1
        return fake_head.next


        



if __name__=='__main__':
    s = Solution()

    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    new_head = s.swapPairs(head)
    print(new_head.val)
    print(new_head.next.val)
    print(new_head.next.next.val)
    print(new_head.next.next.next.val)

    
        